src/mpl2/src/SimulatedAnnealingCore.cpp

```diff
@@ -569,16 +568,28 @@ void SimulatedAnnealingCore<T>::fastSA()
   const float min_t = 1e-10;
   const float t_factor
       = std::exp(std::log(min_t / init_temperature_) / max_num_step_);
-  notch_weight_ = 0.0;  // notch pealty is too expensive, we try to avoid
-                        // calculating notch penalty at very beginning
-  // const for restart
+
+  // Used to ensure notch penalty is used only in the latter steps
+  // as it is too expensive
+  notch_weight_ = 0.0;
+
   int num_restart = 1;
   const int max_num_restart = 2;
-  // SA process
+
+  SequencePair best_valid_result;
+  if (isValid()) {
+    updateBestValidResult(best_valid_result);
+  }
+
   while (step <= max_num_step_) {
     for (int i = 0; i < num_perturb_per_step_; i++) {
       perturb();
       cost = calNormCost();
+
+      if (cost < pre_cost && isValid()) {
+        updateBestValidResult(best_valid_result);
+      }
```

Comment:
Its possible you have no valid result so far and while `cost < pre_cost` is false it is a valid result.  This will not record it.

Reply:
I'll push a change to ensure we also check if the best result is empty.